CONFIG-README.md
# Configuration File

RecolorHelper v0.3 includes a configuration file "recolor.cfg" which can be edited using a standard text editor (notepad++ is recommended).

The default file looks like this:

```
#
#  RecolorHelper config
#  Version 0.3.0 
#  Released June 12 2021
#
#  CONFIG DOCUMENTATION:
#  https://github.com/brongo/recolorHelper/blob/main/CONFIG-README.md
#
#  MAKE CHANGES BELOW
#

color1 = 0.0001;
color2 = 0.00001;
color3 = 0.000001;

useSpecial = false;
specialSearchType = none;

# ALWAYS BACKUP YOUR WORK BEFORE USING THIS
deleteUnmodifiedFiles = false;
```

## Editing the Configuration File

- Lines starting with "#" are comments and will be skipped.
- All lines that are not comments **MUST** end with a semicolon (;) or the program will error.

The variables that can be changed are:

```
color1  ("r" or "x" color values)
color2  ("g" or "y" color values)
coloe3  ("b" or "z" color values)

useSpecial  (set to "true" if using specialSearchType)
specialSearchType  (see below for list of valid options)
deleteUnmodifiedFiles  (set to "true" to delete files that don't contain color values)
```

## Color Values

The colors can be set to specific values. For example:

```
color1 = 1;
color2 = 0;
color3 = 0;
```

This configuration would set all the red channels to 1, while setting green and blue channels to 0. It would make everything bright red.

**Please note the valid range for a color is from 0.0 - 1.0. Any values above 1 will result in an error message.** If you aren't sure what color to use, you can do a Google search for "RGB color picker." Most color pickers will have a value from 0-255 for each channel (r, g, b). To convert this to the format used by the game, simply divide by 255. For example, if the color picker gives you a value of 100. The value used in the game is: `(100/255) = 0.3921568`.

## Ignoring Color Channels

You can opt to ignore (leave unchanged) any color channel by setting the value to `-1`. For example:

```
color1 = 0;
color2 = -1;
color3 = 0;
```
This configuration would set all red and blue channels to zero. It would leave the green channel (color2) unchanged.

## Delete Unmodified Files / Empty Directories

This tool will not delete any files by default. However, you can enable this option in the config by setting:

```
deleteUnmodifiedFiles = true;
```

The name of this setting isn't quite accurate. What is *really* does is delete any files that don't contain color values recognized by the tool. It also will not delete **anything** in the fx/character or fx/demon_player folder, unless using `specialSearchType` settings discussed below. If those values aren't set, the tool skips this folder entirely and will not delete them.

Sometimes a file will not be modified, but also won't be deleted. For example, if you run the program twice with the same settings, *technically* the 2nd pass won't change anything, but it will not delete your files. This is done as a safety measure to prevent accidental deletions.

If using this setting, any empty directories will also be removed.

## Special Search Types

These are used only for .decl files in the **fx/character** and **fx/demon_player** folders. By default, the program will ignore these files. If you want to edit files in these folders, you must set `useSpecial = true;` and set `specialSearchType` to one of the following options.

**Options:**
```
all
bfg_stun
ice_bomb_freeze
on_fire
plasma_stun
```

**Only one option can be used at a time. A working example would be:**

```
useSpecial = true;
specialSearchType = plasma_stun;
```

This configuration tells recolorHelper to **only** modify the colors in fx/character and fx/demon_player folders that are associated with the "plasma_stun" group. The purpose of this option is to allow for recoloring plasma rifle effects independently from other weapon effects such as "on_fire", "bfg_stun", etc. Since those effects will generally be a different color.

**Note:** using `specialSearchType = all;` will change all 4 of the types listed (bfg_stun, ice_bomb_freeze, on_fire, and plasma_stun). It will still ignore other values. Most values outside of these effects should not be changed anyway. If you feel that some option is missing, please send feedback to @SamPT.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [0.3.0] - 2021-06-12

- Added config file (recolor.cfg), for users to set custom config options.
- Added config option for custom color values, set via config file.
- Added config option to ignore color channels (example: set only "red" channels, ignore green and blue).
- Added config option to remove unmodified files.
- Added config options to target specific fx/character and fx/demon_player groups.
- Added support for fx/character and fx/demon_player files.
- Added support for FX_SOUND color used by plasma rifle.
- Added support for "valueVec4" colors used by plasma rifle heat blast mod.
- Additional error handling and safety checks throughout program.
- Significant code refactor for maintainability.
- Windows & Linux codebases are merged. Program can be compiled on either platform.
- Source code is now public and licensed under GPL.

## [0.2.0] - 2021-05-23

- Added support for **FX types:** 
  ```
  1. FX_DECAL
  2. FX_DESTRUCTIBLE
  3. FX_RIBBON_2
  4. FX_SCREEN_SHAKE
  ```
- Added support for **Standard types:** 
  ```
  1. emissiveTint
  2. explosionGradient 
  3. explosionGradient2
  4. explosionGradient3
  5. emissiveMaskColor
  6. emissiveMaskColor2
  7. lightColor
  8. surfaceColor2
  ```
- Added support for **Particle types:** 
  ```
  1. color
  2. fadeColor
  3. initialColor
  4. gradient
  5. explosionGradient
  6. explosionGradient2
  7. explosionGradient3
  8. emissiveMaskColor
  9. emissiveMaskColor2
  ```
- Added multiple error checks and error messages.
- Added small delay before closing program, so user has time to read console output in Windows.
- Fixed a bug that could cause both xyz and rgb values to be printed in some cases.
- Fixed a bug that could cause some FX values to be skipped.

## [0.1.1] - 2021-05-17

- Fixed issue with parsing "editorVars" (unusual .decl formatting).
- Removed .bat file, added short delay before closing console (Windows version).

## [0.1.0] - 2021-05-16

- Initial Release


recolorHelper/Config.h
#pragma once
#include <vector>
#include <string>

namespace RecolorTool
{
    class Config
    {
        public:
            int cfgHeaderSize = 5;
            int maxFileCount = 1000;
            int maxSearchIterationCount = maxFileCount * 2;
            int maxColorLookAhead = 12;
            int maxVarLookAhead = 100; // only used in config files
            int maxVarLookBehind = 5;
            int maxInputPathLength = 512;
            int minInputPathLength = 8;

            std::string colorValue1 = "0.0001";
            std::string colorValue2 = "0.00001";
            std::string colorValue3 = "0.000001";
        
            std::string useSpecial = "false";
            std::string specialSearchType = "none";
            size_t specialSearchTypeIndex = -1;

            std::string deleteUnmodifiedFiles = "false";

            // List of supported color types
            std::vector<std::string> colorTypes
            {
                "type",                 // fxTypes
                "colorOnSpawn",
                "colorOnDeath",
                "decalEmmissiveTint",
                "decalDiffuseTint",
                "emissiveTint",
                "explosionGradient",
                "explosionGradient2",
                "explosionGradient3",
                "emissiveMaskColor",
                "emissiveMaskColor2",
                "lightColor",
                "particleColorTint",
                "surfacecolor2",
                "surfaceemissivecolor",
                "surfaceemissivecolor2",
                "surfacesheencolor",
                "default_noMatch"
            };
            std::vector<std::string> particleColorTypes
            {
                "initialColor",
                "fadeColor",
                "color",
                "explosionGradient",
                "explosionGradient2",
                "explosionGradient3",
                "emissiveMaskColor",
                "emissiveMaskColor2",
                "default_noMatch"
            };
            std::vector<std::string> fxTypes
            {
                "\"FX_LIGHT\"",         // change all except "ambient_light"
                "\"FX_RENDERPARM\"",    // change none except "fire_primary"
                "\"FX_SOUND\"",         // change none except "fire_primary"
                "\"FX_PARTICLE\"",
                "\"FX_MODEL\"",
                "\"FX_SCREEN_SHAKE\"",
                "\"FX_RIBBON_2\"",
                "\"FX_DECAL\"",
                "\"FX_DESTRUCTIBLE\"",
                "default_noMatch"
            };
            std::vector<std::string> valueVec4Types
            {
                "\"CHARGE_PROPERTY_RENDER_PARMS\"",
                "default_noMatch"
            };
            std::vector<std::string> specialTypes
            {
                "\"FX_DECAL\"",
                "\"FX_PARTICLE\"",
                "default_noMatch"
            };

            // List of predefined search types
            std::vector<std::string> predefinedSearchList
            {
                "all",
                "bfg_stun",
                "ice_bomb_freeze",
                "on_fire",
                "plasma_stun",
                "default_noMatch"
            };
            


    };
}
